pandas/tests/indexing/test_indexing.py

```diff
@@ -131,6 +132,8 @@ def test_setitem_dtype_upcast(self):
         assert is_float_dtype(left['foo'])
         assert is_float_dtype(left['baz'])
 
+    @pytest.mark.skipif(PY2, reason=("Catching warnings unreliable with "
```

Comment:
?? where does this come from? this means we are  NOT catching a warning. need to pin this down.

Reply:
Yes, as I wrote: https://github.com/pandas-dev/pandas/pull/20770#pullrequestreview-115735997

A bug in testing code only, which arises with Python 2 only, and only when two test files are run _together_, is in the intersection of my "no idea where to start" and "not high priority anyway" categories.

I can open an issue once merged - as there's no simple way (as far as I know) to reproduce otherwise.

Reply:
As I said, see https://github.com/pandas-dev/pandas/pull/20770#issuecomment-383271735

Reply:
(the full suite case behaves like the case with two files)

Reply:
None of the travis builds are showing any warnings.

I agree with @jreback we should figure out what is going on before merging. But given the tight timeline for 0.23, *if* this would be the only remaining issue, I would rather merge it before the rc and fix it before the final release.

Anyhow, will try to take a closer look at the full PR tomorrow. Quick general comment: I think this deserves a more elaborate whatsnew notice?

Reply:
> if this would be the only remaining issue

I'm afraid I won't have time for (inherited) docstrings and learning what is a ``AbstractClassError`` soon.

> I would make an exception in this case

To be honest I don't see _to what_ is an exception been made. This PR fixes 3 bugs in code, and the fact that it identifies an additional, already existing, one in tests is an added bonus.

Reply:
@toobaz I am not questing the refactoring or that you fixed some bugs, just that something that was added is hiding a bug, maybe in an exception. This is much worse than known bugs and very very hard to find.

Reply:
Do we all agree that the bug is in the test suite? If yes, I guess I don't have much to add: as you correctly point out, it is a nasty bug, which will be hard to isolate - even harder if we don't merge this (because none of us knows how to reproduce otherwise).

Reply:
but the point is that this refactor created the bug. i have no problem waiting to merge this until this is fixed. would rather have this correct, this is pretty tricky code.

Reply:
> but the point is that this refactor created the bug

This "but" has no logical meaning in this discussion. Either you agree with my reasoning that the bug is in the test suite, or you think that a bug in the code can lead a test to pass only when called in isolation, and then I ask you to explain me how.  In the first case, this refactor didn't "create" any bug. In the second... I'm patiently waiting for an explanation (not for one more repetition of the same illogical claim).

Then, sure, there might be a bug in the code _and_ a bug in the tests suite. That's why I have manually tried the code in the failing tests, and it emits the warning just as it should - feel free to try.

The point is not just 0.23.0, it is that we have no idea of what the problem is, it is certainly not _caused_ by this branch, it will be most probably not fixed soon, this branch will soon become a rebasing nightmare (I know by experience), and the indexing code desperately needs cleanup (and it is already difficult enough).

And of course, the point is also that I still didn't see any logical argument for postponing. You know,  I like when discussions are based on actual arguments (i.e. when they are not just a waste of time).